avocado/core/virt.py

```diff
@@ -268,34 +277,54 @@ def setup_login(self, hostname, username, password=None):
         else:
             self.logged = False
 
-    def ip_address(self):
+    def ip_address(self, timeout=30):
         """
-        Returns the IP address associated with the MAC address of this VM
-
-        Inspired from:
-        https://raw.githubusercontent.com/mcepl/virt-addr/master/virt-addr.py
+        Returns the IP address consulting libvirt domifaddr
 
         :returns: either the IP address or None if not found
         :rtype: str or None
         """
-        desc = etree.fromstring(self.domain.XMLDesc(0))
-        mac_path = "devices/interface[@type='network']/mac"
-        node = desc.find(mac_path)
-        if node is None:
-            return None
-
-        mac = node.get("address")
-        if mac is None:
-            return None
-
-        mac = mac.lower().strip()
-        output = subprocess.Popen(["arp", "-n"],
-                                  stdout=subprocess.PIPE).communicate()[0]
-        lines = [line.split() for line in output.split("\n")[1:]]
-        addresses = [line[0] for line in lines if (line and (line[2] == mac))]
-        if addresses:
-            # Just return the first address, this is a best effort attempt
-            return addresses[0]
+        timelimit = time.time() + timeout
+
+        # VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE: Parse DHCP lease file
+        # VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_AGENT: Ask Qemu Guest Agent
+        querytype = libvirt.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_AGENT
+
+        while True:
+            try:
+                ip = self._get_ip(querytype)
+                if ip is not None:
+                    return ip
+            except libvirt.libvirtError as exception:
+                # Qemu guest agent takes time to be ready, but
+                # libvirt raises an exception here if it's not.
+                # Let's be nice and wait for the guest agent, if
+                # that's the problem.
+                errno = libvirt.VIR_ERR_AGENT_UNRESPONSIVE
+                if exception.get_error_code() == errno:
+                    pass
+                else:
+                    return None
+
+            if time.time() > timelimit:
+                return None
+            time.sleep(1)
+
+    def _get_ip(self, querytype):
+        """
+        Retrieves from libvirt the first IP from the
+        first non-loopback device using the selected method.
+
+        :param querytype: libvirt query type code.
+        :return: either the IP address or None if not found.
+        """
+        ifaces = self.domain.interfaceAddresses(querytype)
+        for iface, data in ifaces.iteritems():
+            if data['addrs']:
+                addr = data['addrs'][0]['addr']
```

Comment:
One comment here: this will get either IPv4 or IPv6 addresses, whichever appear first.  libvirt sets IPv4 addresses a `type` of `0`.  We may want to filter for IPv4 only, as I'm not sure fabric behaves well with IPv6 addresses.  But maybe I'm being hysterical about not including IPv6.